widget/diagramwidget/diagram.go

```diff
@@ -0,0 +1,251 @@
+package diagramwidget
+
+import (
+	"image/color"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/driver/desktop"
+	"fyne.io/fyne/v2/theme"
+	"fyne.io/fyne/v2/widget"
+)
+
+var Globaldiagram *DiagramWidget
+
+// ForceRepaint is a workaround for a Fyne bug (Issue #2205) in which moving a canvas object does not
+// trigger repainting. When the issue is resolved, this function and all references to it should be
+// removed. The DummyBox on the GlobalDiagram should also be removed.
+// The conditionals here are required during initialization.
+func ForceRepaint() {
+	if Globaldiagram != nil {
+		if Globaldiagram.DummyBox != nil {
+			Globaldiagram.DummyBox.Refresh()
+		}
+	}
+}
+
+// Verify that interfaces are fully implemented
+var _ fyne.Tappable = (*DiagramWidget)(nil)
+
+type linkPinPair struct {
+	link *DiagramLink
+	pin  *LinkPoint
+}
+
+type DiagramWidget struct {
+	widget.BaseWidget
+
+	// ID is expected to be unique across all DiagramWidgets in the application.
+	ID string
+
+	// Diagrams may want to use a different theme and variant than the application. The default value is the
+	// applicaton's theme and variant
+	DiagramTheme fyne.Theme
+	ThemeVariant fyne.ThemeVariant
+	Offset       fyne.Position
+
+	// DesiredSize specifies the size which the graph widget should take
+	// up, defaults to 800 x 600
+	DesiredSize fyne.Size
```

Comment:
This is an anti-pattern for Fyne widgets.
The desired size is controlled by the application and container it uses for the widget, not from the widget instance itself.